lib/shelley/src/Cardano/Wallet/Shelley/Network/Blockfrost.hs

```diff
@@ -287,24 +400,382 @@ withNetworkLayer tr network np project k =
+                    pure $ bfHeaderHash == toText blockHeaderHash
+
+            getBlockHeaderAtHeight :: Integer -> IO (Maybe BlockHeader)
+            getBlockHeaderAtHeight height =
+                either (error . show) Just . fromBlockfrost <$>
+                    (runBF . BF.getBlock) (Left height)
+
+            getBlockHeaderAt :: ChainPoint -> IO (Maybe BlockHeader)
+            getBlockHeaderAt = runBF . \case
+                ChainPointAtGenesis -> do
+                    block <- BF.getBlock $ Right $ genesisBlockHash network
+                    pure . either (error . show) Just $ fromBlockfrost block
+                ChainPoint (SlotNo slot) blockHeaderHash -> do
+                    b@BF.Block {_blockHash = BF.BlockHash bfHeaderHash} <-
+                        BF.getBlockSlot (BF.Slot (toInteger slot))
+                    pure $ if bfHeaderHash == toText blockHeaderHash
+                        then either (error . show) Just $ fromBlockfrost b
+                        else Nothing
+
+            getNextBlocks :: ChainPoint -> IO (Maybe [Block])
+            getNextBlocks = runBF . \case
+                ChainPointAtGenesis ->
+                    Just . pure <$> fetchGenesisBlock network
+                ChainPoint _slotNo hash ->
+                    -- Only one block is fetched for now, even though the type
+                    -- allows for a list of blocks
+                    Just . pure <$> fetchBlock network hash
+
+            getAddressTxs
+                :: BlockHeader
+                -> BlockHeader
+                -> Either Address RewardAccount
+                -> IO ChainEvents
+            getAddressTxs bhFrom bhTo = runBF . fmap fromBlockEvents . \case
+                Left address -> do
+                    txs <- BF.allPages \paged -> BF.getAddressTransactions'
+                        (BF.Address (encodeAddress @nd address))
+                        paged
+                        BF.Ascending
+                        (Just $ headerToIndex bhFrom)
+                        (Just $ headerToIndex bhTo)
+                    forConcurrently txs \BF.AddressTransaction{..} -> do
+                        (bftx, tx) <-
+                            fetchTransaction network _addressTransactionTxHash
+                        txIndex <-
+                            _addressTransactionTxIndex
+                                <?#> "_addressTransactionTxIndex"
+                        txBlockEvents
+                            bftx
+                            (unsafeMkSublist [(txIndex, tx)])
+                            (wholeList [])
+                Right account -> do
+                    let address = BF.Address $ encodeStakeAddress @nd account
+                    rs <- BF.getAccountRegistrations address <&> fmap \reg ->
+                        ( BF._accountRegistrationTxHash reg
+                        , pure case BF._accountRegistrationAction reg of
```

Comment:
Style: Hrm, I don't know. I'm more a fan of

```suggestion
                        , pure $ case BF._accountRegistrationAction reg of
```

because that would make it visually more clear that `pure` has only one argument instead of four — the four arguments being `case`, `BF._accountRegistrationAction`, `reg` and `of`. 😆 Yes, yes, I know that `case` is a keyword, but it takes me a moment to parse that visually. (Similar reason for why I'm not too fond of `BlockArguments`.)